From 3440f8d0f6d9d84e146101fddfcdc388b3777630 Mon Sep 17 00:00:00 2001
From: Maruthi Srinivas Bayyavarapu <maruthi.srinivas.bayyavarapu@xilinx.com>
Date: Thu, 6 Sep 2018 17:43:31 +0530
Subject: [PATCH 0115/1851] ASoC: xlnx: add PL sound card device creation

commit e50e1b967faebd00735a536939d085114e8e8ffc from
https://github.com/Xilinx/linux-xlnx.git

Sound card device is a logical device for which there is no device tree
node. This device is created if any of the source,sink node is found in
parsing of audio formatter device tree node. These source/sink nodes will
be used in sound card creation.

Signed-off-by: Maruthi Srinivas Bayyavarapu <maruthi.srinivas.bayyavarapu@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 sound/soc/xilinx/xlnx_formatter_pcm.c | 43 ++++++++++++++++++++++++---
 sound/soc/xilinx/xlnx_snd_common.h    | 17 +++++++++++
 2 files changed, 56 insertions(+), 4 deletions(-)
 create mode 100644 sound/soc/xilinx/xlnx_snd_common.h

diff --git a/sound/soc/xilinx/xlnx_formatter_pcm.c b/sound/soc/xilinx/xlnx_formatter_pcm.c
index 46019df15811..3d7bd8e9321d 100644
--- a/sound/soc/xilinx/xlnx_formatter_pcm.c
+++ b/sound/soc/xilinx/xlnx_formatter_pcm.c
@@ -16,6 +16,8 @@
 #include <sound/soc.h>
 #include <sound/pcm_params.h>
 
+#include "xlnx_snd_common.h"
+
 #define DRV_NAME "xlnx_formatter_pcm"
 
 #define XLNX_S2MM_OFFSET	0
@@ -87,6 +89,7 @@ struct xlnx_pcm_drv_data {
 	int mm2s_irq;
 	struct snd_pcm_substream *play_stream;
 	struct snd_pcm_substream *capture_stream;
+	struct platform_device *pdev;
 };
 
 /*
@@ -347,10 +350,10 @@ static int xlnx_formatter_pcm_trigger(struct snd_pcm_substream *substream,
 
 static int xlnx_formatter_pcm_new(struct snd_soc_pcm_runtime *rtd)
 {
-	struct snd_card *card = rtd->card->snd_card;
-
+	struct snd_soc_component *component = snd_soc_rtdcom_lookup(rtd,
+								    DRV_NAME);
 	snd_pcm_lib_preallocate_pages_for_all(rtd->pcm,
-			SNDRV_DMA_TYPE_DEV, card->dev,
+			SNDRV_DMA_TYPE_DEV, component->dev,
 			xlnx_pcm_hardware.buffer_bytes_max,
 			xlnx_pcm_hardware.buffer_bytes_max);
 	return 0;
@@ -378,6 +381,8 @@ static int xlnx_formatter_pcm_probe(struct platform_device *pdev)
 	u32 val;
 	struct xlnx_pcm_drv_data *aud_drv_data;
 	struct resource *res;
+	struct device_node *nodes[XLNX_MAX_PATHS];
+	struct device *dev = &pdev->dev;
 
 	aud_drv_data = devm_kzalloc(&pdev->dev,
 				    sizeof(*aud_drv_data), GFP_KERNEL);
@@ -413,6 +418,16 @@ static int xlnx_formatter_pcm_probe(struct platform_device *pdev)
 			return ret;
 		}
 		xlnx_formatter_pcm_reset(aud_drv_data->mmio + XLNX_MM2S_OFFSET);
+
+		nodes[XLNX_PLAYBACK] = of_parse_phandle(dev->of_node,
+							"xlnx,tx", 0);
+		if (!nodes[XLNX_PLAYBACK])
+			dev_err(&pdev->dev, "tx node not found\n");
+		else
+			dev_info(&pdev->dev,
+				 "sound card device will use DAI link: %s\n",
+				 (nodes[XLNX_PLAYBACK])->name);
+		of_node_put(nodes[XLNX_PLAYBACK]);
 	}
 	if (val & AUD_CFG_S2MM_MASK) {
 		aud_drv_data->s2mm_presence = true;
@@ -430,6 +445,16 @@ static int xlnx_formatter_pcm_probe(struct platform_device *pdev)
 			return ret;
 		}
 		xlnx_formatter_pcm_reset(aud_drv_data->mmio + XLNX_S2MM_OFFSET);
+
+		nodes[XLNX_CAPTURE] = of_parse_phandle(dev->of_node,
+						       "xlnx,rx", 0);
+		if (!nodes[XLNX_CAPTURE])
+			dev_err(&pdev->dev, "rx node not found\n");
+		else
+			dev_info(&pdev->dev,
+				 "sound card device will use DAI link: %s\n",
+				 (nodes[XLNX_CAPTURE])->name);
+		of_node_put(nodes[XLNX_CAPTURE]);
 	}
 
 	dev_set_drvdata(&pdev->dev, aud_drv_data);
@@ -441,8 +466,16 @@ static int xlnx_formatter_pcm_probe(struct platform_device *pdev)
 		return ret;
 	}
 
-	dev_info(&pdev->dev, "pcm platform device registered\n");
+	if (nodes[XLNX_PLAYBACK] || nodes[XLNX_CAPTURE])
+		aud_drv_data->pdev =
+			platform_device_register_resndata(&pdev->dev,
+							  "xlnx_snd_card", 0,
+							  NULL, 0, &nodes,
+							  sizeof(nodes));
+	if (!aud_drv_data->pdev)
+		dev_err(&pdev->dev, "sound card device creation failed\n");
 
+	dev_info(&pdev->dev, "pcm platform device registered\n");
 	return ret;
 }
 
@@ -450,6 +483,8 @@ static int xlnx_formatter_pcm_remove(struct platform_device *pdev)
 {
 	struct xlnx_pcm_drv_data *adata = dev_get_drvdata(&pdev->dev);
 
+	platform_device_unregister(adata->pdev);
+
 	if (adata->s2mm_presence)
 		xlnx_formatter_pcm_reset(adata->mmio + XLNX_S2MM_OFFSET);
 
diff --git a/sound/soc/xilinx/xlnx_snd_common.h b/sound/soc/xilinx/xlnx_snd_common.h
new file mode 100644
index 000000000000..ab3c82a0e122
--- /dev/null
+++ b/sound/soc/xilinx/xlnx_snd_common.h
@@ -0,0 +1,17 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Xilinx ASoC sound card support
+ *
+ * Copyright (C) 2018 Xilinx, Inc.
+ */
+
+#ifndef _XLNX_SND_COMMON_H
+#define _XLNX_SND_COMMON_H
+
+enum {
+	XLNX_PLAYBACK,
+	XLNX_CAPTURE,
+	XLNX_MAX_PATHS
+};
+
+#endif /* _XLNX_SND_COMMON_H */
-- 
2.31.1

